### PR TITLE
feat: ll-hls warnings and info

### PR DESCRIPTION
**Pull request by brandonocasey**
1. Adds warning/info logs for all ll-hls tags
2. Adds defaults for ll-hls tags

### CODE REVIEW

File: src/parser.js [hunk 1]
```diff
@@ -6,6 +6,55 @@ import decodeB64ToUint8Array from '@videojs/vhs-utils/es/decode-b64-to-uint8-arr
 import LineStream from './line-stream';
 import ParseStream from './parse-stream';
 
+// set SERVER-CONTROL hold back based upon targetDuration and partTargetDuration
+// we need this helper because defaults are based upon targetDuration and
+// partTargetDuration being set, but they may not be if SERVER-CONTROL appears before
+// target durations are set.
+const setHoldBack = function(manifest) {
```
Comment (brandonocasey) [inline src/parser.js:13]: Had to break this into it's own function see the comment above this function.

File: src/parser.js [hunk 2]
```diff
@@ -386,24 +437,106 @@ export default class Parser extends Stream {
             },
             'skip'() {
               this.manifest.skip = entry.attributes;
+
+              if (!entry.attributes.hasOwnProperty('SKIPPED-SEGMENTS')) {
+                this.trigger('warn', {
```
Comment (brandonocasey) [inline src/parser.js:442]: Should we trigger an error for things like this??

File: test/fixtures/m3u8/llhlsDelta.json [hunk 3]
```diff
@@ -172,7 +173,7 @@
     "CAN-BLOCK-RELOAD": true,
     "CAN-SKIP-UNTIL": 12,
     "PART-HOLD-BACK": 1,
-    "HOLD-BACK": 2
+    "HOLD-BACK": 12
```
Comment (brandonocasey) [inline test/fixtures/m3u8/llhlsDelta.json:176]: targetDuration is 4 for this source, which means HOLD-BACK has to be at least 12.

File: src/parser.js [hunk 4]
```diff
@@ -6,6 +6,55 @@ import decodeB64ToUint8Array from '@videojs/vhs-utils/es/decode-b64-to-uint8-arr
...
+    this.trigger('warn', {
+      message: `${tag} clamping ${hb} (${serverControl[hb]}) to targetDuration * 3 (${minTargetDuration})`
+    });
+    serverControl[hb] = minTargetDuration;
+  }
+
+  // default no part hold back to part target duration * 3
+  if (partTargetDuration && !serverControl.hasOwnProperty(phb)) {
+    serverControl[phb] = partTargetDuration * 3;
+    this.trigger('info', {
```
Comment (gkatsev) [inline src/parser.js:43]: should we have this as `warn` if we have an `EXT-X-PART-INF` tag in the manifest?

File: src/parser.js (hunk 2 above)
Comment (gkatsev) [reply src/parser.js:442]: As long as triggering an error won't cause us to stop parsing or something, we should, otherwise, I think we can continue with a warning.

File: src/parser.js [hunk 5]
```diff
@@ -386,24 +437,114 @@ export default class Parser extends Stream {
...
+                });
+              }
             },
             'part'() {
               this.manifest.parts = this.manifest.parts || [];
               this.manifest.parts.push(entry.attributes);
+              const missingAttributes = [];
+
+              ['URI', 'DURATION'].forEach(function(k) {
+                if (!entry.attributes[k]) {
```
Comment (gkatsev) [inline src/parser.js:453]: if a duration is zero, this could theoretically fail, no? Might be better to use `k in entry.attributes`?

File: src/parser.js [hunk 6]
```diff
@@ -386,24 +437,114 @@ export default class Parser extends Stream {
...
+                  missingAttributes.push(k);
+                }
+              });
+
+              if (missingAttributes.length) {
+                const index = this.manifest.parts.length - 1;
+
+                this.trigger('warn', {
+                  message: `#EXT-X-PART #${index} lacks required attribute(s) ${missingAttributes.join(', ')}`
```
Comment (gkatsev) [inline src/parser.js:462]:
```suggestion
                  message: `#EXT-X-PART #${index} lacks required attribute(s): ${missingAttributes.join(', ')}`
```

File: src/parser.js [hunk 7]
```diff
@@ -386,24 +437,114 @@ export default class Parser extends Stream {
...
+                });
+              }
+
+              if (this.manifest.renditionReports) {
+                this.manifest.renditionReports.forEach((r, i) => {
+                  if (!r.hasOwnProperty('LAST-PART')) {
+                    this.trigger('warn', {
+                      message: `#EXT-X-RENDITION-REPORT #${i} lacks required attribute(s) LAST-PART`
```
Comment (gkatsev) [inline src/parser.js:470]: so, this is only required if the manifest has a partial segment. Can we inspect here to see if we got one or not? I can imagine that it's possible to have a rendition report without partial segments.

File: src/parser.js [hunk 8]
```diff
@@ -386,24 +437,114 @@ export default class Parser extends Stream {
...
+              this.manifest.serverControl = attrs;
+              if (!attrs.hasOwnProperty('CAN-BLOCK-RELOAD')) {
+                this.manifest.serverControl['CAN-BLOCK-RELOAD'] = false;
+                this.trigger('info', {
+                  message: '#EXT-X-SERVER-CONTROL defaulting CAN-BLOCK-RELOAD to false'
+                });
+              }
+              setHoldBack.call(this, this.manifest);
+
+              if (attrs.hasOwnProperty('CAN-SKIP-DATERANGES') && !attrs.hasOwnProperty('CAN-SKIP-UNTIL')) {
```
Comment (gkatsev) [inline src/parser.js:488]: are these props only set if their values are YES?

File: src/parser.js [hunk 9]
```diff
@@ -386,24 +437,114 @@ export default class Parser extends Stream {
...
+                if (!entry.attributes[k]) {
+                  missingAttributes.push(k);
+                }
+              });
+
+              if (missingAttributes.length) {
+                const index = this.manifest.preloadHints.length - 1;
+
+                this.trigger('warn', {
+                  message: `#EXT-X-PRELOAD-HINT #${index} lacks required attribute(s) ${missingAttributes.join(', ')}`
```
Comment (gkatsev) [inline src/parser.js:511]:
```suggestion
                  message: `#EXT-X-PRELOAD-HINT #${index} lacks required attribute(s): ${missingAttributes.join(', ')}`
```

File: src/parser.js [hunk 10]
```diff
@@ -386,24 +437,114 @@ export default class Parser extends Stream {
...
             },
             'rendition-report'() {
               this.manifest.renditionReports = this.manifest.renditionReports || [];
               this.manifest.renditionReports.push(entry.attributes);
+              const index = this.manifest.renditionReports.length - 1;
+              const missingAttributes = [];
+              const warning = `#EXT-X-RENDITION-REPORT #${index} lacks required attribute(s)`;
+
+              ['LAST-MSN', 'URI'].forEach(function(k) {
+                if (!entry.attributes[k]) {
```
Comment (gkatsev) [inline src/parser.js:523]: here as well, may be better to use `k in entry.attributes` as LAST-MSN is a number and could theoretically be zero.

File: test/parser.test.js [hunk 11]
```diff
@@ -383,31 +383,424 @@ QUnit.module('m3u8s', function(hooks) {
...
+      this.infos = [];
 
-  QUnit.test('triggers warning for missing EXT-X-START TIME-OFFSET attribute', function(assert) {
+      this.parser.on('warn', (warn) => this.warnings.push(warn.message));
+      this.parser.on('info', (info) => this.infos.push(info.message));
 
-    const manifest = [
+    }
+  });
+  QUnit.test('warn when #EXT-X-TOTAL-DURATION is invalid', function(assert) {
```
Comment (gkatsev) [inline test/parser.test.js:396]:
```suggestion
  QUnit.test('warn when #EXT-X-TARGETDURATION is invalid', function(assert) {
```

File: test/parser.test.js [hunk 12]
```diff
@@ -383,31 +383,424 @@ QUnit.module('m3u8s', function(hooks) {
...
+      '#EXT-X-DISCONTINUITY-SEQUENCE:0',
+      '#EXT-X-TARGETDURATION:foo',
+      '#EXTINF:10,',
+      'media-00001.ts',
+      '#EXT-X-ENDLIST'
+    ].join('\n'));
+    this.parser.end();
+
+    const warnings = [
+      'ignoring invalid target duration: undefined'
```
Comment (gkatsev) [inline test/parser.test.js:409]: this has `undefined` because we tried to parse `foo` as an int which failed, and we just didn't set a value in that case?

If we would want to fix this, we'd want to do it in a separate PR.

File: src/parser.js (hunk 4 above)
Comment (brandonocasey) [reply src/parser.js:43]: It's valid for it not to exist, the spec just specifies a default to use.

File: src/parser.js (hunk 7 above)
Comment (brandonocasey) [reply src/parser.js:470]: This is within the part handler, just incase the first rendition report happens before the first part.

File: src/parser.js (hunk 8 above)
Comment (brandonocasey) [reply src/parser.js:488]: I think this should be a boolean check here, as we only expect CAN-SKIP-UNTIL if `CAN-SKIP-DATERANGES` is `YES` (which will be true here already.

File: src/parser.js (hunk 2 above)
Comment (brandonocasey) [reply src/parser.js:442]: I think we should continue with a warning unless we find out that this stops us from playing a source back.

File: src/parser.js (hunk 4 above)
Comment (gkatsev) [reply src/parser.js:43]: The spec says it's required if if PART-INF is present (end of part-hold-back section in https://tools.ietf.org/html/draft-pantos-hls-rfc8216bis-08#section-4.4.3.8)
>       PART-HOLD-BACK is REQUIRED if the Playlist contains the EXT-X-PART-INF tag.

Comment (brandonocasey) [reply src/parser.js:43]: ahh I see the confusion here now. partTargetDuration maps to `PART-TARGET` the only attribute of `EXT-X-PART-INF` we can warn if we have a `manifest.partInf` without a `PART-TARGET`, but we already do that for `EXT-X-PART-INF` and it seems like we would be logging the same thing twice.

File: test/parser.test.js (hunk 12 above)
Comment (brandonocasey) [reply test/parser.test.js:409]: I think I will be fixing up some of the old warnings and info logs to better represent what we are doing now in another pull request.

File: src/parser.js (hunk 2 above)
Comment (gkatsev) [reply src/parser.js:442]: Sounds good.

**Review by gkatsev — APPROVED** [review]
